Add bookings

alekseytumbaev opened the pull request.

alekseytumbaev committed:
style:
- Add curly braces in every if statement
- In userMapper. Write arguments from new line in User constructor
refactor:
- Introduce constant for 'X-Sharer-User-Id' header
- add set of emails in InMemoryUserStorage to check if email is unique

alekseytumbaev committed:
feat:
- Connection to postgres (properties and data-jpa)
- javax.persistence annotations for Item and User

alekseytumbaev committed:
feat:
- UserRepository, change UserService to use repo
- properties for data schema generation

alekseytumbaev committed:
feat:
- ItemRepository
- @ManyToOne at owner Item.owner field
- ItemMapper now retrieves owner (using UserService) when converting dto to Item
delete:
- EmailAlreadyExistsException, because uniqueness is checked by database
- Storage interfaces and InMemoryStorage implementations, because of conflicts with changed code.

alekseytumbaev committed:
feat:
- add booking endpoints
- add change status dto

alekseytumbaev committed:
feat:
- add item's last and next bookings

alekseytumbaev committed:
feat:
- endpoint for creating comments
fix:
- ItemService.getItemNextBooking() doesn't return rejected and canceled bookings anymore
style:
- rename methods and classes with names *DtoResponse to *ResponseDto

alekseytumbaev committed:
feat:
- add comments to ItemWithBookingsResponseDto
fix:
- ItemService.getItemLastBooking() now returns booking with start time before current time, not end time before current;
- BookingRepository now doesn't check status of Booking when finding by CURRENT state

alekseytumbaev committed:
feat:
- add schema.sql
remove:
- some annotations from jpa entities

alekseytumbaev committed:
turn off dev profile with postgres db

alekseytumbaev committed:
style:
- code style correction

chehovnp commented on src/main/java/ru/practicum/shareit/booking/exception/UnknownBookingState.java:
```diff
@@ -0,0 +1,13 @@
+
+import lombok.Getter;
+
+public class UnknownBookingState extends RuntimeException {
```
название указывает на некий статус,  нужно именовать с использованием слова Exception

chehovnp commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,91 @@
+@Validated
 public class BookingController {
+    private final BookingService bookingService;
+    private final BookingMapper bookingMapper;
```
оставь тут только сервис ,  bookingMapper перенеси в сам сервис. 
второй момент касается самого маппера , напишу там

chehovnp commented on src/main/java/ru/practicum/shareit/booking/model/BookingMapper.java:
```diff
@@ -0,0 +1,62 @@
+
+@Component
+public class BookingMapper {
+    private final ItemService itemService;
```
маппер это утилитный класс,  он трансформирует Dto в сущность и наоборот.
сделай его методы статическими ,  объект BookingMapper инжектить не нужно будет . 
сервисы от сюда убери . далее ниже напишу

chehovnp commented on src/main/java/ru/practicum/shareit/booking/model/BookingMapper.java:
```diff
@@ -0,0 +1,62 @@
+
+    public Booking toBooking(BookingRequestDto bookingRequestDto, long bookingId, long bookerId, BookingStatus bookingStatus) {
+        Item item = itemService.getById(bookingRequestDto.getItemId());
+        User user = userService.getById(bookerId);
```
эти страшные методы перенеси в сервис. 
там есть транзакционность - управление запросами в базу.

chehovnp commented on src/main/java/ru/practicum/shareit/booking/model/BookingMapper.java:
```diff
@@ -0,0 +1,62 @@
+                booking.getStart(),
+                booking.getEnd(),
+                userMapper.toUserDto(booking.getBooker()),
+                itemMapper.toItemDto(booking.getItem()),
```
можем же в сервисе эти методы                 
userMapper.toUserDto(booking.getBooker()),
itemMapper.toItemDto(booking.getItem()),
вызвать , 
в метод toBookingResponseDto передать Booking и два результата DTO

chehovnp commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,91 @@
+            @RequestParam(defaultValue = "ALL") String state) {
+        BookingState bookingState = convertToBookingState(state);
+
+        Collection<Booking> bookings = bookingService.getAllForUserItemsSortedByStartTimeDesc(itemOwnerId, bookingState);
```
верни из bookingService.getAllForUserItemsSortedByStartTimeDesc(itemOwnerId, bookingState); уже Collection<BookingResponseDto>

chehovnp commented on src/main/java/ru/practicum/shareit/item/model/ItemMapper.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public Item toItem(ItemDto itemDto) {
+        User owner = userService.getById(itemDto.getOwnerId());
```
как и в прошлом маппере ,  выноси от сюда сервис

chehovnp commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -1,68 +1,210 @@
+
+        List<ItemWithBookingsResponseDto> itemDtos = new ArrayList<>();
+        for (Item item : items) {
+            List<CommentResponseDto> commentResponseDtos = commentRepo.findByItem_Id(item.getId())
```
1 запрос getAllByOwnerId, посчитай пожалуйста , сколько будет запросов в базу , если мы получим 1000 items. 

- хочу чтобы ты именно написал ответом. 
- второе,  если больше трех, то нужно переделать:
1 запрос. items = itemRepo.findAllByOwner_Id(ownerId)
2 запрос. commentRepo.findByItem_Id(items)
3 запрос. bookingRepo.findAllByItem_Id(items))
желательно получить не List<CommentResponseDto> а Map<Long,List<CommentResponseDto>>   itemId против List<CommentResponseDto> 
далее собрать воедино и вернуть.

alekseytumbaev replied:
Переделал. Теперь отправляются только три запроса:
```java
public Collection<ItemWithBookingsResponseDto> getAllByOwnerId(long ownerId) {
    Collection<Item> items = itemRepo.findAllByOwner_Id(ownerId);

    List<Long> itemIds = items.stream().map(Item::getId).collect(Collectors.toList());
    Map<Long, List<Comment>> itemIdToComments = commentRepo.findAllByItem_IdAsMap(itemIds);
    Map<Long, List<Booking>> itemIdToBookings = bookingRepo.findAllByItem_IdAsMap(itemIds);
    ...
}
```

chehovnp replied:
вижу. 
огонь, молодец! горжусь тобой!

chehovnp reviewed (commented):
Алексей,  из серьездного getAllByOwnerId!!
это у тебя там баг находится- запросы в цикле, еще называют ошибка N+1

alekseytumbaev committed:
refactor:
- make all methods in mappers static
- mappers now used only in services
- rename UnknownBookingState to UnknownBookingStateException

alekseytumbaev committed:
refactor:
- add CommentRepository.findAllByItem_IdAsMap and BookingRepository.findAllByItem_IdAsMap to retrieve comments and bookings for item ids in one query
